src/node_quic_session.cc

```diff
@@ -154,64 +98,21 @@ int KeyCB(
     void* arg) {
   QuicSession* session = static_cast<QuicSession*>(arg);
 
-  if (session->OnKey(name, secret, secretlen) != 0)
-    return 0;
-
-  return 1;
-}
-
-// NOTE(@jasnell): The majority of this is adapted directly from the
-// example code in https://github.com/ngtcp2/ngtcp2. It can likely
-// use a refactor to be more Node-ish
-
-// inspired by <http://blog.korfuri.fr/post/go-defer-in-cpp/>, but our
-// template can take functions returning other than void.
-template <typename F, typename... T> struct Defer {
-  Defer(F &&f, T &&... t)
-      : f(std::bind(std::forward<F>(f), std::forward<T>(t)...)) {}
-  Defer(Defer &&o) noexcept : f(std::move(o.f)) {}
-  ~Defer() { f(); }
-
-  using ResultType = typename std::result_of<typename std::decay<F>::type(
-      typename std::decay<T>::type...)>::type;
-  std::function<ResultType()> f;
-};
-
-template <typename F, typename... T> Defer<F, T...> defer(F &&f, T &&... t) {
-  return Defer<F, T...>(std::forward<F>(f), std::forward<T>(t)...);
-}
-
-inline void prf_sha256(CryptoContext* ctx) { ctx->prf = EVP_sha256(); }
-
-inline void aead_aes_128_gcm(CryptoContext* ctx) {
-  ctx->aead = EVP_aes_128_gcm();
-  ctx->hp = EVP_aes_128_ctr();
-}
-
-inline size_t aead_key_length(const CryptoContext& ctx) {
-  return EVP_CIPHER_key_length(ctx.aead);
-}
-
-inline size_t aead_nonce_length(const CryptoContext& ctx) {
-  return EVP_CIPHER_iv_length(ctx.aead);
+  return session->OnKey(name, secret, secretlen) != 0 ? 0 : 1;
 }
 
-inline size_t aead_tag_length(const CryptoContext &ctx) {
-  if (ctx.aead == EVP_aes_128_gcm() || ctx.aead == EVP_aes_256_gcm()) {
-    return EVP_GCM_TLS_TAG_LEN;
-  }
-  if (ctx.aead == EVP_chacha20_poly1305()) {
-    return EVP_CHACHAPOLY_TLS_TAG_LEN;
-  }
-  UNREACHABLE();
-}
+#define RETURN_IF_FAIL(test, success, ret) if (test != success) return ret;
+#define RETURN_IF_FAIL_OPENSSL(test) RETURN_IF_FAIL(test, 1, -1)
```

Comment:
I’d guess you’re disagreeing, but I don’t think these really help with readability/accessibity of the code all that much…

If you do keep them, please add parentheses around the expanded variables here and wrap them in a `do { ... } while(0)`, for example

```c++
#define RETURN_IF_FAIL(test, success, ret) do {  \
      if ((test) != (success)) return (ret);     \
    } while(0)
#define RETURN_IF_FAIL_OPENSSL(test) RETURN_IF_FAIL((test), 1, -1)
```